src/algorithms/classification/extra_trees.py

```diff
@@ -0,0 +1,291 @@
+        label_str = f'[label=<{pivot_str}<br/>{samples_str}<br/>{class_str}>, fillcolor="{color_hex}", shape=box]'
+        return label_str
+
+
+class DecisionTree:
+
+    def __init__(self, data: pd.DataFrame, label_name: str, features_number: int, min_child_number: int, max_depth: int,
+                 min_metrics: float = 0, metrics_type: metrics_types = 'gini'):
+        self.data = data
+        self.label_name = label_name
+        self.features_number = features_number
+        self.min_child_number = min_child_number
+        self.max_depth = max_depth
+        self.min_metrics = min_metrics
+        self.metrics_type = metrics_type
+        self.features = pd.Series(data.columns.drop(self.label_name))
+        self.root = self.create_node(pd.Series([True] * len(data), index=data.index), 1)
+
+    def create_node(self, mask: pd.Series, depth) -> Node | Leaf:
+        if not self._can_be_split(mask, depth):
+            return Leaf(self.data[self.label_name][mask])
+        features = self.features.sample(self.features_number)
+        best_metrics = None
+        left_mask = None
+        right_mask = None
+        best_feature = None
+        best_pivot = None
+        for feature in features:
+            values = pd.Series(list(set(self.data[feature][mask])))
+            if check_numeric(self.data[feature]):
+                pivot = values.sample(1).iloc[0]
+            else:
+                if len(values) <= 2:
+                    pivot = values.sample(1).iloc[0]
+                else:
+                    n = np.random.randint(1, len(values))
+                    pivot = values.sample(n)
+            result = self._split(mask, feature, pivot)
+            if isinstance(result, bool):
+                continue
+            metrics, left, right = result
+            if best_metrics is None or metrics > best_metrics:
+                best_metrics = metrics
+                left_mask = left
+                right_mask = right
+                best_feature = feature
+                best_pivot = pivot
+
+        if best_metrics is None:
+            return Leaf(self.data[self.label_name][mask])
+
+        node = Node(best_feature, best_pivot)
+        node.importance = best_metrics
+        node.left = self.create_node(left_mask, depth + 1)
+        node.right = self.create_node(right_mask, depth + 1)
+        node.samples = np.sum(mask)
+        node.largest_class = self.data[self.label_name][mask].value_counts().index[0]
+        return node
+
+    def _can_be_split(self, mask: pd.Series, depth: int) -> bool:
+        if len(set(self.data[self.label_name][mask])) == 1: # pure node
+            return False
+        if self.max_depth is not None and self.max_depth <= depth:
+            return False
+        if np.sum(mask) < 2 * self.min_child_number: # or add parameter
+            return False
+        return True
+
+    def _split(self, mask: pd.Series, label: str, pivot: any) -> bool | Tuple[float, pd.Series, pd.Series]:
```

Comment:
I think ... -> Optional[Tuple...] may be more consistent

Reply:
But it will require replace handling False with None, so if you want, you can keep actual version.